### ISSUE TITLE
xCAT management of kvm host

### ISSUE THREAD

**Issue by AndyDevman** [xcat2/xcat-core#6751]
Hi All,

I have recently installed an xCAT management node on CentOSv8. - The node is a virtual machine within an vSphere cluster.

I have manually created another CentOS VM within the vSphere cluster and have installed KVM for it to run as a hypervisor.
On this same virtual machine, I have also installed VirtualBMC to simulate BMC and enable control using the ipmitool-xcat tool. 

My aim is to use xCAT to manage the virtualized KVM hypervisor , enabling creation and management of virtual machines.

I was wondering if anyone had already set up a similar environment or would be able to share any tips / pointers / guides as unfortunately I seem to be drawing a blank.

Thanks

Andy

**Comment by gurevichmark** [comment]
Have you tried defining your virtualized KVM hypervisor as a regular node with `mgt=ipmi` ?
Then when create VMs managed by this virtualized KVM hypervisor, set their `vmhost=` to point it and `mgt=kvm` ?

**Comment by AndyDevman** [comment]
Hi Mark,

thanks for your reply and suggestion.  

I've tried to set the mgt flag for the hypervisor as ipmi, but without any success yet.
I am unsure what the username/password should be set to as I dont think vmware guests support IPMI by default ? perhaps I could use a local account which has sudo privilege ?

Perhaps the kvm hypervisor can only be managed using 'mgt=esx'  - as it is virtual machine running within our vmware vsphere cluster. ?

**Comment by gurevichmark** [comment]
@AndyDevman Since you mentioned that you have installed VirtualBMC on this virtualized KVM hypervisor, I thought that `mgt=ipmi` would work.

But if you only care about using the virtualized KVM hypervisor for creation and management of VMs, and not about management of the virtualized KVM hypervisor itself, maybe the `mgt=` for it, is not that important.
Since you already have your hypervisor setup and running, try to define and create VMs on it using these instructions:
https://xcat-docs.readthedocs.io/en/stable/guides/admin-guides/manage_clusters/ppc64le/virtual_machines/manage_vms.html

**Comment by AndyDevman** [comment]
@gurevichmark , thanks for your quick reply.

I've had a go at setting things up in this way but don't seem to be having much luck and not certain where I'm going wrong
Apologies in advance for the long post and if my questions are obvious.

When running the mkvm command against vm1, everything seems to complete without any errors however nothing happens on the kvm Hypervisor.
Network between the control node and the kvm hypervisor seems OK and ssh keys are in place etc.

Not certain if this helps, but here is the output of the tables which have been modified as found from tabgrep

**nodelist**
---------
#node,groups,status,statustime,appstatus,appstatustime,primarysn,hidden,updatestatus,updatestatustime,zonename,comments,disable
"KVM-hypervisor","all,kvm",,,,,,,,,,,
"vm1","vm,all",,,,,,,,,,,


**hypervisor**
----------
#node,type,mgr,interface,netmap,defaultnet,cluster,datacenter,preferdirect,comments,disable
"KVM-hypervisor","kvm",,,,,,,,,


**hosts**
------
#node,ip,hostnames,otherinterfaces,comments,disable
"KVM-hypervisor","192.168.2.16",,,,
"vm1","192.168.0.4",,,,


**nodetype**
----------
#node,os,arch,profile,provmethod,supportedarchs,nodetype,comments,disable
"KVM-hypervisor","centos",,,,,,,


**nodehm**
----------
#node,power,mgt,cons,termserver,termport,conserver,serialport,serialspeed,serialflow,getmac,cmdmapping,consoleondemand,consoleenabled,comments,disable
"KVM-hypervisor",,"ipmi",,,,,,,,,,,,,
"vm1",,"kvm",,,,,,,,,,,,,


**vm**
----
#node,mgr,host,migrationdest,storage,storagemodel,storagecache,storageformat,cfgstore,memory,cpus,nics,nicmodel,bootorder,clockoffset,virtflags,master,vncport,textconsole,powerstate,beacon,datacenter,cluster,guestostype,othersettings,physlots,vidmodel,vidproto,vidpassword,comments,disable
"vm1",,"KVM-hypervisor",,,,,,,"512","1","virbr0","virtio",,,,,,,,,,,,,,,,,,


**noderes**
--------
#node,servicenode,netboot,tftpserver,tftpdir,nfsserver,monserver,nfsdir,installnic,primarynic,discoverynics,cmdinterface,xcatmaster,current_osimage,next_osimage,nimserver,routenames,nameservers,proxydhcp,syslog,comments,disable
"vm1",,"xnba",,,,,,,,,,,,,,,,,,,


**mac** - redacted
----
#node,interface,mac,comments,disable
"control-node",,"12:12:12:12:12:12",,
"vm1",,"12:12:12:12:12:12",,



I was wondering if the passwd table should be any different ? Should the root user on the kvm host match the system/root user in the xcat passwd table ?  I wasnt sure if I could just define a host as a the key and then add any user and password so long as they existed on the node which you are trying to interact with ?
How should the xcat control node interact & communicate with the kvm hypervisor ?


**passwd** 
---------
#key,username,password,cryptmethod,authdomain,comments,disable
"system","root","abc123",,,"default root user for compute nodes",
"omapi","xcat_key","RANDOM-NUMBER-HERE",,,,
"ipmi","admin","password",,,"default IPMI user/password for VirtualBMC",
"KVM-hypervisor","test-user","Password",,,,

**Comment by gurevichmark** [comment]
Can you show the output of these 2 commands:
`lsdef KVM-hypervisor`, `lsdef vm1` and `lsvm KVM-hypervisor`

**Comment by AndyDevman** [comment]
Hi @gurevichmark , thanks for your help with this.  
Here are the outputs from the commands below

**lsdef vm1**
Object name: vm1
    groups=vm,all
    ip=192.168.0.4
    mac=42:44:c0:a8:00:04
    mgt=kvm
    netboot=xnba
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles
    vmcpus=1
    vmhost=KVM-hypervisor
    vmmemory=512
    vmnicnicmodel=virtio
    vmnics=virbr0


**lsdef KVM-hypervisor**
Object name: KVM-hypervisor
    groups=all,kvm
    hosttype=kvm
    ip=192.168.2.16
    os=centos
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles


**lsvm KVM-hypervisor**
KVM-hypervisor: [control-node]: Error: Unable to identify plugin for this command, check relevant tables: hypervisor.type;nodetype.os=(esx.*);nodehm.mgt;nodehm.mgt=ipmi;nodehm.mgt=kvm;hypervisor.type=(rhev.*)

**Comment by gurevichmark** [comment]
Try adding `mgt=ipmi` to `KVM-hypervisor` definition and running `lsvm KVM-hypervisor` again.

Also, are you able to do passwordless ssh from the xCAT management node to the KVM-hypervisor ?

**Comment by AndyDevman** [comment]
Hi @gurevichmark . 

I've added the mgt=ipmi to the KVM-hypervisor, in the nodehm table so the most recent output from lsdef is as follows

**lsdef KVM-hypervisor**
Object name: KVM-hypervisor
    groups=all,kvm
    hosttype=kvm
    ip=192.168.2.16
    mgt=ipmi
    os=centos
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles


The output from lsvm is now as follows.  
**lsvm KVM-hypervisor**
KVM-hypervisor: Could not get any information about specified object


I've included the output of ping to show that basic connectivity to the node seems OK.
**PING** KVM-hypervisor (192.168.2.16) 56(84) bytes of data.
64 bytes from KVM-hypervisor (192.168.2.16): icmp_seq=1 ttl=64 time=0.223 ms
64 bytes from KVM-hypervisor (192.168.2.16): icmp_seq=2 ttl=64 time=0.173 ms
64 bytes from KVM-hypervisor (192.168.2.16): icmp_seq=3 ttl=64 time=0.190 ms
64 bytes from KVM-hypervisor (192.168.2.16): icmp_seq=4 ttl=64 time=0.157 ms

**Comment by gurevichmark** [comment]
@AndyDevman 
In the code, we try to do this to get the connection to the hypervisor:
```$hypconn = Sys::Virt->new(uri => "qemu+ssh://root@" . $_ . "/system?no_tty=1");```
I suspect a passwordless ssh to the root on hypervisor is needed.

**Comment by AndyDevman** [comment]
Hi @gurevichmark ,

I have just checked the hypervisor and can see the relevant ssh keys are in place
When logged in as root on the xcat control node, I can successfully ssh root@KVM-hypervisor without being prompted for a password.  I have just run ssh root@192.168.2.16 so that the IP is also in the know_hosts file.

Interestingly, when following the logs on the virtualization host and simultaneously running the mkvm command against the vm1 node definition, the following sequence of messages appear

systemd[1]: Started Session 22 of user root
systemd-logind[1142]: Session 22 logged out. Waiting for processes to exit.
systemd-logind[1142]: Removed session 22.

**Comment by gurevichmark** [comment]
Not sure what those messages mean.
You could add some print statements to `/opt/xcat/lib/perl/xCAT_plugin/kvm.pm` and see where it breaks. I think functions `sub dohyp` and `sub mkvm` would be most relevant.

**Comment by AndyDevman** [comment]
I think the messages are related to openssh and indicate that connections are happening when the mkvm command is run.
I have added some print statements to the suggested functions within the kvm module, but not sure where the console will output to ?  Should I be using another module to output to /var/log/messages ?

Thanks
Andy

**Comment by gurevichmark** [comment]
@AndyDevman Sorry, I should have been more explicit. When I said "print statement", I meant using `xCAT::SvrUtils::sendmsg()`, there are several examples of those in `kvm.pm`

**Comment by AndyDevman** [comment]
Hi @gurevichmark.  That's no problem thanks very much for the explanation.  
I added some print statements to the mkvm function as suggested and worked out that it wasn't getting as far as the storage sections.   I set the vmstorage field to a location which is local to the hypervisor and ran the mkvm with the '-s' flag.
At this stage I could see that some of the if statements further down the function were being accessed.
Looking at the hypervisor , I can now see that the disk has been created in the location specified as follows
vm1.sda.qcow2
rpower vm1 on now works too ! and I can verify that the VM has appeared as running on the hypervisor when running virsh.
Thanks very much for your help troubleshooting this one.
